rfcs/0008-partial-backfill-ccw.md

```diff
@@ -0,0 +1,408 @@
+* It is reasonably clear how the feature would be implemented.
+* Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+-->
+
+The detailed design is TBD;
+  I want to think this one through "in code", first, via a prototype or three.
+
+One thing I'm explicitly _not_ doing is shying away from using this challenge
+  as a place to explore how we might also meet other current and upcoming architectural needs.
+This is an intentional decision: we **need** to start making progress on some of these needs/challenges
+  and this effort is as good an excuse as any.
+This additional exploration work will need to be balanced against the **more urgent**
+  need to actually _ship_ the partial backfill solution, though.
+
+
+#### Design Option A: Hacky Approach
+
+I'm just mentioning this one for completeness' sake, as it is not suitable to a recurring need like this will be
+  (we may need to run a partial backfill quarterly, going forwards).
+That said, one approach is to just build a custom version of the BFD Pipeline where:
+
+* most fields are `Optional`
+* most of the tests are removed (as they will fail to compile given the `Optional` change)
+* the BFD Server is not built
+* `RifLoader` was configured to **only** run `_MERGE` data sets
+
+The BFD team would also have to carefully orchestrate things to ensure
+  that this custom build was produced and deployed at the correct time,
+  and then reverted/undeployed back to the normal BFD Pipeline after the backfill was complete.
+
+Why is this approach a bad idea?
+Well, as mentioned, it's very manual and would have to be repeated from scratch every quarter.
+Being so manual, it's extremely error-prone, which is bad by itself,
+  but given how painful a DB restore would be, is just far too risky.
+
+
+#### Design Option B: Overhaul the Existing Code
+
+The second-most minimal approach would be to update the existing code to add a couple new layers to the ETL:
+
+1. Create separate structs/classes to represent parsed RIF records.
+2. Add a RIF-to-JPA entity mapping layer in that supports both partial and full loads.
+
+That's a brief description for some incredibly complex work.
+It'd require extensiive changes and extensions to our automatic source code generation,
+  which is by far the trickiest bit of code we have.
+
+There are a number of risks with this approach:
+
+* It's a lot of tedious and tricky work.
+    * This is true of all our options, but still worth mentioning.
+* It risks some performance degradation,
+    as moving data through those extra layers will require a lot of additional memory copies, GC pressure, etc.
+* Keeping the changes current/merged with the main branch in Git will be very difficult.
+
+That said, it's not a _bad_ approach if the backfill is the only thing we want to try and address.
+
+
+#### Design Option C: Create a New Pipeline
```

Comment:
@karlmdavis I'm interested in better understanding the approach you're proposing for `Design Option C`; I'm going to attempt to paraphrase to see if I understand this fully.

This option hinges around creating a contract between the application that loads data into the BFD database and the application(s) that extract data out of the upstream datastores (ie CCW and eventually RDA API).  

The extraction application(s), possibly making use of a framework like Airflow, AWS Glue or just plain lambda functions, would implement different processes for different data types.  Extract that data triggered on S3 and/or cron events, and publish normalized messages into a messaging stream (eg kafka or kinesis).

Within BFD, a new idempotent (thinking about duplicate messages) process would consume the messages out of the stream and write the changes to the postgres db.

At a high level, is that what you're proposing?

Reply:
This was addressed in the unresolved questions section of the RFC